src/chocolatey/infrastructure.app/nuget/ChocolateyNugetCredentialProvider.cs

```diff
@@ -86,67 +86,62 @@ public Task<CredentialResponse> GetAsync(Uri uri, IWebProxy proxy, CredentialReq
             }
 
             // credentials were not explicit
-            // discover based on closest match in sources
-            var candidateSources = _config.MachineSources.Where(
-                s =>
-                {
-                    var sourceUrl = s.Key.TrimEnd('/');
-
-                    try
-                    {
-                        var sourceUri = new Uri(sourceUrl);
-                        return sourceUri.Host.IsEqualTo(uri.Host)
-                            && !string.IsNullOrWhiteSpace(s.Username)
-                            && !string.IsNullOrWhiteSpace(s.EncryptedPassword);
-                    }
-                    catch (Exception)
-                    {
-                        this.Log().Error("Source '{0}' is not a valid Uri".FormatWith(sourceUrl));
-                    }
-
-                    return false;
-                }).ToList();
-
+            // find matching source(s) in sources list
+            var trimmedTargetUri = new Uri(uri.AbsoluteUri.TrimEnd('/'));
             MachineSourceConfiguration source = null;
 
-
-            if (candidateSources.Count == 1)
+            // If the user has specified --source with a *named* source and not a URL, try to find the matching one
+            // with the correct URL for this credential request.
+            var namedExplicitSources = _config.ExplicitSources?.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries)
+                .Where(s => !Uri.IsWellFormedUriString(s, UriKind.Absolute))
+                .ToList();
+            if (namedExplicitSources?.Count > 0)
             {
-                // only one match, use it
-                source = candidateSources.FirstOrDefault();
+                // Uri.Equals() and == operator compare hostnames case-insensitively and the remainder of the url case-sensitively
+                // while ignoring #fragments on the URLs, but does care about trailing slashes, which we do not here.
+                source = _config.MachineSources
+                    .Where(s => namedExplicitSources.Contains(s.Name) && new Uri(s.Key.TrimEnd('/')) == trimmedTargetUri)
+                    .FirstOrDefault();
             }
-            else if (candidateSources.Count > 1)
+
+            if (source is null)
             {
-                // find the source that is the closest match
-                foreach (var candidateSource in candidateSources.OrEmpty())
+                // Could not find a valid source by name, or the source(s) specified were all URLs.
+                // Try to look up the target URL in the saved machine sources to attempt to match credentials.
+                //
+                // Note: This behaviour remains as removing it would be a breaking change, but we may want
+                // to remove this in a future version, as specifying an explicit URL should potentially

```

Comment:
If there has been a decision to remove this, or potentially removing it, we would need an issue to be created to keep track of this.

Reply:
I think we agree that we would add an issue to track this discussion, and point to it in this comment in the code, right?

Reply:
🤔 I'm not sure how this would be deprecated in any meaningful way, honestly? It's kind of an edge case implementation detail, but. Yeah we can hammer that out in the issue when we get to it. 👍

Reply:
Cory already raised it this morning looks like, so added some comments to it. https://github.com/chocolatey/choco/issues/3573

Reply:
> 🤔 I'm not sure how this would be deprecated in any meaningful way, honestly? It's kind of an edge case implementation detail, but. Yeah we can hammer that out in the issue when we get to it. 👍

What I meant by deprecating it, would probably be to output a message to the user that the functionality of loading settings for absolute URLs will be removed in the future. (Assuming that is the decision that will be made).

> Cory already raised it this morning looks like, so added some comments to it. https://github.com/chocolatey/choco/issues/3573

👍